File: 450-DSA-SHEET/ARRAYS/Python/9-Minimize_the_Heights_II.py
import math
from numpy import np

class Solution:
    def getMinDiff(self,arr,n,k):
        arr.sort()
        ans = arr[n-1]-arr[0]
        for i in range(1,n):
            if (k<=arr[i]):
                maxE = max(arr[i-1]+k, arr[n-1]-k)
                minE = min(arr[0]+k, arr[i]-k)
                ans = min(ans, (maxE - minE))
        return ans


if __name__ == "__main__":
    tc = int(input())
    while(tc > 0):
        k = int(input())
        n = int(input())
        arr = list(map(int, input().strip(),spli
        ))
        ob = Solution()
        ans = ob.getMinDiff(arr,n,k)
        print(ans)
        tc-=1

File: 450-DSA-SHEET/ARRAYS/C-Plus-Plus/9-Minimize_the_Heights_II.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
        int getMinDiff(int arr[], int n, int k){
            sort(arr,arr+n);
            int maxE, minE;
            int ans = arr[n-1] - arr[0];
            for(int i=1;i<n;i++){
                if(k <= arr[i]){
                    maxE = max(arr[i-1]+k, arr[n-1]-k);
                    minE = min(arr[0]+k, arr[i]-k);
                    ans = min(ans, (maxE-minE));
                }
            }
            return ans;
        }
};


int main(){
    int t; cin>>t;
    while(t--){
        int n,k; cin>>n>>k;
        int arr[n];
        for(int i=0;i<n;i++){
            cin>>arr[i];
        }
        Solution ob;
        auto ans = ob.getMinDiff(arr,n,k);
        cout << ans << endl;
    }
    return 0;
}